StoreKitService

gMock-28 opened the pull request:
StoreKitService + ProductStatusStorage

gMock-28 committed:
StoreKitService

StoreKitService + ProductStatusStorage

k0zl0v commented on Mokay/Mokay/StoreKit/ProductStatusStorage.swift:
```diff
@@ -0,0 +1,49 @@
+
+import Foundation
+
+public final class ProductStatusStorage {
```
нужен конформ Sendable

k0zl0v commented on Mokay/Mokay/StoreKit/ProductStatusStorage.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    // MARK: - Private methods
+    
+    private func loadProductStatuses() -> [String: Bool] {
+        return userDefaults.dictionary(forKey: productStatusKey) as? [String: Bool] ?? [:]
+    }
```
Лучше завести модель данных вместо словаря, а то не очень масшатбируемо

k0zl0v commented on Mokay/Mokay/StoreKit/StoreKitService.swift:
```diff
@@ -0,0 +1,117 @@
+import Combine
+import StoreKit
+
+@MainActor
+public final class StoreKitService: ObservableObject {
```
убрать MainActor и Observable

k0zl0v replied:
Пометить Sendable

k0zl0v commented on Mokay/Mokay/StoreKit/StoreKitService.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    // MARK: - Public properties
+    
+    @Published var products: [Product] = []
```
Переделать на CurrentValueSubject или AsyncStream

gMock-28 replied:
Вынес CurrentValueSubject в актор ProductStorage

k0zl0v commented on Mokay/Mokay/StoreKit/StoreKitService.swift:
```diff
@@ -0,0 +1,117 @@
+    // MARK: - Private properties
+    
+    private let productStatusStorage: ProductStatusStorage
+    private let productIds: [String]
```
Перенести в стор

k0zl0v commented on Mokay/Mokay/StoreKit/StoreKitService.swift:
```diff
@@ -0,0 +1,117 @@
+    private let productStatusStorage: ProductStatusStorage
+    private let productIds: [String]
+    
+    private var updateTask: Task<Void, Never>? = nil
```
для чего хранить таск?

k0zl0v commented on Mokay/Mokay/StoreKit/StoreKitService.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    // MARK: - Public methods
+    
+    public func fetchProducts() async throws {
```
можно сделать приватной и вызвать в ините

k0zl0v commented on Mokay/Mokay/StoreKit/StoreKitService.swift:
```diff
@@ -0,0 +1,117 @@
+                    throw error
+                }
+                
+            case .pending:
+                break
```
получается функция выполнится успешно. С точки зрения использования сервиса немного странно получается, что мы как бы сообщаем об успешной покупке, хотя она находится в pending

k0zl0v commented on Mokay/Mokay/StoreKit/StoreKitService.swift:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+    
+    public func updatePurchasedProducts() async {
```
должна быть приватной

k0zl0v commented on Mokay/Mokay/StoreKit/ProductStatusStorage.swift:
```diff
@@ -0,0 +1,49 @@
+    // MARK: - Private properties
+    
+    private let userDefaults: UserDefaults
+    private let productStatusKey = "ProductStatusKey"
```
ключи должны быть в формате com.Mokay.ProductStatusKey

gMock-28 committed:
StoreKitService

Исправления после ревью:
- StoreKitService и ProductStorage подписаны под Sendable/@unchecked Sendable
- Изменена модель хранения продуктов в UserDefaults
- Получение списка продуктов с помощью AsyncStream
- Хранение списка id продуктов перенесено в ProductStorage
- Описаны возможные ошибки, возвращаемые StoreKitService, в StoreKitSericeError

Commit:
StoreKitService

Вынес CurrentValueCubject в актор ProductStorage
